### ISSUE TITLE
beautify default column names

### ISSUE THREAD

**Issue by xudong963** [apache/arrow-datafusion#2027]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Currently, DataFusion creates column names by https://github.com/apache/arrow-datafusion/blob/7be0e268a69ffecbf06823c98ca572733dddb29e/datafusion/src/physical_plan/planner.rs#L91.  
The approach has two existing problems:
1. result in potential bugs, such as https://github.com/apache/arrow-datafusion/issues/1818.
2. Makes the printed table unattractive and may be very wide, such as
```sql
❯ select COALESCE(null, null, null, null);
+-------------------------------------------------------+
| coalesce(Utf8(NULL),Utf8(NULL),Utf8(NULL),Utf8(NULL)) |
+-------------------------------------------------------+
|                                                       |
+-------------------------------------------------------+
1 row in set. Query took 0.002 seconds.
```
**Describe the solution you'd like**
1. Simplify default column names by modifying https://github.com/apache/arrow-datafusion/blob/7be0e268a69ffecbf06823c98ca572733dddb29e/datafusion/src/physical_plan/planner.rs#L91.
2. Directly define a default column name, such as `?column?` in postgres:
```sql
postgres=# select 1, 2;
 ?column? | ?column?
----------+----------
        1 |        2
(1 row)
```

**Comment by WinkerDu** [comment]
I would like to pick this up. Please assign to me as appropriate @xudong963

**Comment by xudong963** [comment]
Thank you @WinkerDu !

**Comment by yjshen** [comment]
Please also consider the Apache Hive auto alias as a reference. 

I'm conservative on the need to provide a default name. If something goes wrong, it's acceptable to let the query fail, requiring the user to alias explicitly.

**Comment by WinkerDu** [comment]
@yjshen  After trying the following 2 cases in Hive 2.3.7, I got the outputs below:
```
hive> set hive.cli.print.header=true;
hive> select 1,2;
OK
_c0	_c1
1	2
Time taken: 0.272 seconds, Fetched: 1 row(s)
hive> SELECT coalesce(NULL, 1, NULL);
OK
_c0
1
Time taken: 0.228 seconds, Fetched: 1 row(s)
```
In my opinion, Hive auto alias convention `_c0` seems more reasonable since that indicates column index.
I prefer doing this in Datafusion. 
What do you think? @xudong963 @yjshen

**Comment by WinkerDu** [comment]
> Please also consider the Apache Hive auto alias as a reference.
> 
> TBH, I'm conservative on the need to provide a default name. If something goes wrong, it's acceptable to fail the query, requiring the user to alias explicitly.

@xudong963 After checking #1818 , realigning anonymous columns for all `Expr` seems a bit much.
I prefer what @alamb put that adding a `LogicalPlan::Projection` to `Union`. BTW, I'd like to solve that `Union` corner bug

**Comment by xudong963** [comment]
@WinkerDu  Yeah, you can try to solve #1818 first, we can talk more about #2027

**Comment by alamb** [comment]
FWIW I think @houqp  put a substantial amount of thought into the output name semantics of columns, trying to find a balance between doing what an existing system did and being more useful than `?column?`. The details are written in https://arrow.apache.org/datafusion/specification/output-field-name-semantic.html

**Comment by gaoxinge** [comment]
@alamb Is there any specifications for cast expression as column name? 

I raise a [pr](https://github.com/apache/arrow-datafusion/pull/5170) for cast expression, but it not pass all the tests, and I find my solution is ugly is some cases:

```sql
> select cast('1969-01-01T00:00:00.1Z' as timestamp);
"+---------------------------------------------------------------------+",
"| CAST(Utf8(\"1969-01-01T00:00:00.1Z\") AS Timestamp(Nanosecond, None)) |",
"+---------------------------------------------------------------------+",
"| 1969-01-01T00:00:00.100                                             |",
"+---------------------------------------------------------------------+",
```

**Comment by alamb** [comment]
> @alamb Is there any specifications for cast expression as column name?


I do not know of one, sadly. I think what we have done in the past is to look at other sql implementations (like posgres and spark) for inspiration and followed one of their models

**Comment by gaoxinge** [comment]
@alamb Thx. I raise an issue https://github.com/apache/arrow-datafusion/issues/5174 to talk with this case.

**Comment by NevroHelios** [comment]
Since it is still open can I work on it and submit a pr?

**Comment by jayzhan211** [comment]
https://github.com/apache/datafusion/issues/10274 is also a nice issue to work on with.
